_posts/2024-08-10-integration-for-kw-ssh.md

```diff
@@ -0,0 +1,112 @@
+
+![Picture](/images/kw-ssh-illustration.png)
+
+This image illustrates the structure of the integration tests for the `kw-ssh`
+feature, using containers to simulate different operating system environments.
+This setup involves one container acting as the test environment, within which
+tests are executed across three Linux distributions: **Debian**, **Fedora**,
+and **Archlinux**.
+
+Within this test environment, there is a second container, represented in the
+image as "nested" which hosts the SSH server needed for the tests. This
+configuration allows for the isolation of the test environment and execution of
+`kw-ssh` commands on a simulated SSH server, without affecting the local system
+or other containers.
+
+By using containers for each Linux environment and for the SSH server, we
+ensure that tests are conducted in controlled environments, avoiding
+contamination between tests and maintaining result consistency. This approach
+allows the functionality of `kw-ssh` to be validated across different
+distributions, ensuring the code performs as expected on various platforms.
+
+# Details of the Testing Environment
+
+Inside the container that serves as the test environment, I copy a file from
+the host machine, which is the `Containerfile` responsible for generating the
+container with the SSH server. This SSH container is essential for enabling
+connection tests using `kw-ssh`, ensuring that the authentication and transfer
+processes work correctly.
+
+![Picture](/images/containerfile_ssh.png)
+
+# Challenges with Nested Containers
+
+When creating containers within containers, executing commands directly from
+the host machine to the nested container becomes a challenge. To address this
+complexity, I developed the `container_exec_in_nested_container()` function,
+which facilitates command execution within this nested environment.
+
+![Picture](/images/container_nested.png)
+
+This function facilitates the execution of commands in a nested container
+within another container, which is common in kw-ssh integration tests. It uses
+the `container_exec()` function, which executes commands directly in a
+container. One of the challenges when passing commands to containers is
+handling special characters, such as single quotes, which can cause issues
+during execution. To address this, I used the `str_escape_single_quotes()`
+function, which correctly escapes these characters, ensuring that commands are
+executed reliably.
+
+# Managing Commands in Nested Containers
+
+The `str_escape_single_quotes()` function uses the **sed** command to add
+backslashes `(\)` before any single quotes found in the string, allowing
+commands containing single quotes to be interpreted correctly by the shell:
+
+![Picture](/images/str_escape_single_quotes.png)
+
+Additionally, in the `container_exec_in_nested_container()` function, I use the
+special `$''` format for strings, known as "ANSI C quoting". This format allows
```

Comment:
Maybe link to the ANSI-C quoting section in Bash reference manual

https://www.gnu.org/software/bash/manual/bash.html#ANSI_002dC-Quoting